File: apps/web/src/lib/api/health.ts
/**
 * Health API Client
 *
 * Fetches health and metrics data from the backend.
 */

import { fetchWithRetry } from './fetchWithRetry';

const API_BASE = import.meta.env.VITE_API_BASE || '/api';

// Types

export type NetworkType = 'rodalies' | 'metro' | 'bus' | 'tram' | 'fgc';

export type FreshnessStatus = 'fresh' | 'stale' | 'unavailable';

export type HealthStatus = 'healthy' | 'degraded' | 'unhealthy' | 'unknown';

export type OverallStatus = 'operational' | 'degraded' | 'outage';

export type ConfidenceLevel = 'high' | 'medium' | 'low';

export interface DataFreshness {
  network: NetworkType;
  lastPolledAt: string | null;
  ageSeconds: number;
  status: FreshnessStatus;
  vehicleCount: number;
}

export interface NetworkHealth {
  network: NetworkType;
  healthScore: number;
  status: HealthStatus;
  dataFreshness: number;
  serviceLevel: number;
  dataQuality: number;
  vehicleCount: number;
  expectedCount?: number;
  lastUpdated: string;
  confidenceLevel: ConfidenceLevel;
  activeAnomalies: number;
}

export interface OverallHealth {
  status: OverallStatus;
  healthScore: number;
  networks: NetworkHealth[];
  lastUpdated: string;
  uptimePercent: number;
  activeIncidents: number;
}

export interface DataFreshnessResponse {
  networks: DataFreshness[];
  lastChecked: string;
}

export interface NetworkHealthResponse {
  overall: OverallHealth;
  networks: NetworkHealth[];
}

// API Functions

/**
 * Fetch data freshness for all networks
 */
export async function fetchDataFreshness(): Promise<DataFreshnessResponse> {
  const response = await fetchWithRetry(`${API_BASE}/health/data`, {
    logPrefix: 'Health API',
    timeoutMs: 5000,
    useCircuitBreaker: false, // Don't circuit break health checks
  });

  if (!response.ok) {
    throw new Error(`Failed to fetch data freshness: ${response.status}`);
  }

  return response.json();
}

/**
 * Fetch network health scores
 */
export async function fetchNetworkHealth(): Promise<NetworkHealthResponse> {
  const response = await fetchWithRetry(`${API_BASE}/health/networks`, {
    logPrefix: 'Health API',
    timeoutMs: 5000,
    useCircuitBreaker: false,
  });

  if (!response.ok) {
    throw new Error(`Failed to fetch network health: ${response.status}`);
  }

  return response.json();
}

// Utility Functions

/**
 * Get display name for a network
 */
export function getNetworkDisplayName(network: NetworkType): string {
  const names: Record<NetworkType, string> = {
    rodalies: 'Rodalies',
    metro: 'Metro',
    bus: 'Bus',
    tram: 'Tram',
    fgc: 'FGC',
  };
  return names[network] || network;
}

/**
 * Get color for health status
 */
export function getHealthStatusColor(status: HealthStatus): string {
  const colors: Record<HealthStatus, string> = {
    healthy: '#22c55e',    // green-500
    degraded: '#eab308',   // yellow-500
    unhealthy: '#ef4444',  // red-500
    unknown: '#6b7280',    // gray-500
  };
  return colors[status] || colors.unknown;
}

/**
 * Get color for overall status
 */
export function getOverallStatusColor(status: OverallStatus): string {
  const colors: Record<OverallStatus, string> = {
    operational: '#22c55e',
    degraded: '#eab308',
    outage: '#ef4444',
  };
  return colors[status] || '#6b7280';
}

/**
 * Get color for freshness status
 */
export function getFreshnessStatusColor(status: FreshnessStatus): string {
  const colors: Record<FreshnessStatus, string> = {
    fresh: '#22c55e',
    stale: '#eab308',
    unavailable: '#ef4444',
  };
  return colors[status] || '#6b7280';
}

/**
 * Format age in human-readable format
 */
export function formatAge(ageSeconds: number): string {
  if (ageSeconds < 0) return 'Unknown';
  if (ageSeconds < 60) return `${ageSeconds}s ago`;
  if (ageSeconds < 3600) return `${Math.floor(ageSeconds / 60)}m ago`;
  return `${Math.floor(ageSeconds / 3600)}h ago`;
}

File: apps/api/models/health.go
package models

import "time"

// NetworkType represents a transit network
type NetworkType string

const (
	NetworkRodalies NetworkType = "rodalies"
	NetworkMetro    NetworkType = "metro"
	NetworkBus      NetworkType = "bus"
	NetworkTram     NetworkType = "tram"
	NetworkFGC      NetworkType = "fgc"
)

// AllNetworks returns all supported network types
func AllNetworks() []NetworkType {
	return []NetworkType{
		NetworkRodalies,
		NetworkMetro,
		NetworkBus,
		NetworkTram,
		NetworkFGC,
	}
}

// DataFreshness represents the freshness status of data for a network
type DataFreshness struct {
	Network      NetworkType `json:"network"`
	LastPolledAt *time.Time  `json:"lastPolledAt"`
	AgeSeconds   int         `json:"ageSeconds"`
	Status       string      `json:"status"` // "fresh", "stale", "unavailable"
	VehicleCount int         `json:"vehicleCount"`
}

// NetworkHealth represents the health status of a transit network
type NetworkHealth struct {
	Network           NetworkType `json:"network"`
	HealthScore       int         `json:"healthScore"`       // 0-100
	Status            string      `json:"status"`            // "healthy", "degraded", "unhealthy", "unknown"
	DataFreshness     int         `json:"dataFreshness"`     // 0-100 score
	ServiceLevel      int         `json:"serviceLevel"`      // 0-100 score (actual vs expected)
	DataQuality       int         `json:"dataQuality"`       // 0-100 score
	VehicleCount      int         `json:"vehicleCount"`
	ExpectedCount     *int        `json:"expectedCount,omitempty"`
	LastUpdated       time.Time   `json:"lastUpdated"`
	ConfidenceLevel   string      `json:"confidenceLevel"`   // "high", "medium", "low"
	ActiveAnomalies   int         `json:"activeAnomalies"`
}

// OverallHealth represents the overall system health
type OverallHealth struct {
	Status          string          `json:"status"`          // "operational", "degraded", "outage"
	HealthScore     int             `json:"healthScore"`     // 0-100
	Networks        []NetworkHealth `json:"networks"`
	LastUpdated     time.Time       `json:"lastUpdated"`
	UptimePercent   float64         `json:"uptimePercent"`   // Last 24h
	ActiveIncidents int             `json:"activeIncidents"`
}

// AnomalyEvent represents a detected anomaly
type AnomalyEvent struct {
	ID            int64       `json:"id"`
	DetectedAt    time.Time   `json:"detectedAt"`
	Network       NetworkType `json:"network"`
	AnomalyType   string      `json:"anomalyType"`   // "low_vehicle_count", "stale_data", "api_failure"
	Severity      string      `json:"severity"`      // "info", "warning", "critical"
	ExpectedValue *float64    `json:"expectedValue,omitempty"`
	ActualValue   *float64    `json:"actualValue,omitempty"`
	ZScore        *float64    `json:"zScore,omitempty"`
	Description   string      `json:"description"`
	ResolvedAt    *time.Time  `json:"resolvedAt,omitempty"`
	IsActive      bool        `json:"isActive"`
}

// NetworkBaseline represents expected vehicle counts for a network
type NetworkBaseline struct {
	Network            NetworkType `json:"network"`
	HourOfDay          int         `json:"hourOfDay"`
	DayOfWeek          int         `json:"dayOfWeek"`
	VehicleCountMean   float64     `json:"vehicleCountMean"`
	VehicleCountStdDev float64     `json:"vehicleCountStdDev"`
	SampleCount        int         `json:"sampleCount"`
}

// HealthStatus constants
const (
	StatusHealthy     = "healthy"
	StatusDegraded    = "degraded"
	StatusUnhealthy   = "unhealthy"
	StatusUnknown     = "unknown"
	StatusOperational = "operational"
	StatusOutage      = "outage"
)

// FreshnessStatus constants
const (
	FreshnessFresh       = "fresh"       // < 60s
	FreshnessStale       = "stale"       // 60s - 5min
	FreshnessUnavailable = "unavailable" // > 5min or no data
)

// CalculateFreshnessStatus returns the freshness status based on age
func CalculateFreshnessStatus(ageSeconds int) string {
	if ageSeconds < 0 {
		return FreshnessUnavailable
	}
	if ageSeconds < 60 {
		return FreshnessFresh
	}
	if ageSeconds < 300 {
		return FreshnessStale
	}
	return FreshnessUnavailable
}

// CalculateFreshnessScore returns a 0-100 score based on data age
func CalculateFreshnessScore(ageSeconds int) int {
	if ageSeconds < 0 {
		return 0
	}
	if ageSeconds <= 30 {
		return 100
	}
	if ageSeconds >= 300 {
		return 0
	}
	// Linear decay from 100 at 30s to 0 at 300s
	return 100 - ((ageSeconds - 30) * 100 / 270)
}

// CalculateHealthStatus returns health status based on score
func CalculateHealthStatus(score int) string {
	if score >= 80 {
		return StatusHealthy
	}
	if score >= 50 {
		return StatusDegraded
	}
	if score > 0 {
		return StatusUnhealthy
	}
	return StatusUnknown
}
